### PR TITLE
refactor: Convert Sched to Kotlin 

### PR DESCRIPTION
**Pull request by david-allison**

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 1]
```diff
@@ -16,591 +16,544 @@
...
-        }
+        c = _getNewCard()
+        return c ?: _getLrnCard(true)
         // collapse or finish
-        return _getLrnCard(true);
     }
 
-
-    protected @NonNull CardQueue<? extends Card.Cache>[] _fillNextCard() {
+    override fun _fillNextCard(): Array<CardQueue<out Card.Cache?>> {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:241]: @KotlinCleanup("simplify fun with when")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 2]
```diff
@@ -16,591 +16,544 @@
...
-
-    @Override
-    protected void _resetLrnCount() {
-        _resetLrnCount(null);
+    override fun _resetLrnCount() {
+        _resetLrnCount(null)
     }
-    protected void _resetLrnCount(@Nullable CancelListener cancelListener) {
+
+    override fun _resetLrnCount(cancelListener: CancelListener?) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:277]: @KotlinCleanup("see if these functions can be combined into one")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 3]
```diff
@@ -609,78 +562,85 @@ public class Sched extends SchedV2 {
...
-        _resetRevCount(null);
-    }
-    protected void _resetRevCount(@Nullable CancelListener cancelListener) {
-        mRevCount = _walkingCount(d -> _deckRevLimitSingle(d, true),
-                                  this::_cntFnRev, cancelListener);
+    override fun _resetRevCount() {
+        _resetRevCount(null)
     }
 
+    override fun _resetRevCount(cancelListener: CancelListener?) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:599]: @KotlinCleanup("see if these functions can be combined into one")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 4]
```diff
@@ -690,585 +650,546 @@ public class Sched extends SchedV2 {
...
+        val fct = card.factor / 1000.0
+        val ivl2 =
+            _constrainedIvl(((card.ivl + delay / 4) * 1.2).toInt(), conf, card.ivl.toDouble())
+        val ivl3 = _constrainedIvl(((card.ivl + delay / 2) * fct).toInt(), conf, ivl2.toDouble())
+        val ivl4 = _constrainedIvl(
+            ((card.ivl + delay) * fct * conf.getDouble("ease4")).toInt(),
+            conf,
+            ivl3.toDouble()
+        )
         if (ease == Consts.BUTTON_TWO) {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:785]: @KotlinCleanup("simplify with when")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 5]
```diff
@@ -690,585 +650,546 @@ public class Sched extends SchedV2 {
...
         }
-
     }
 
-    @SuppressWarnings("PMD.UnusedFormalParameter") // it's unused upstream as well
-    private int _adjRevIvl(@NonNull Card card, int idealIvl) {
-        idealIvl = _fuzzedIvl(idealIvl);
-        return idealIvl;
+    // it's unused upstream as well
+    private fun _adjRevIvl(idealIvl: Int): Int {
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:816]: @KotlinCleanup("simplify fun")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 6]
```diff
@@ -690,585 +650,546 @@ public class Sched extends SchedV2 {
...
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONArray delays = conf.optJSONArray("delays");
+        val oconf = col.decks.confForDid(card.oDid)
+        var delays = conf.optJSONArray("delays")
         if (delays == null) {
-            delays = oconf.getJSONObject("new").getJSONArray("delays");
+            delays = oconf.getJSONObject("new").getJSONArray("delays")
         }
-        JSONObject dict = new JSONObject();
+        val dict = JSONObject()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:958]: @KotlinCleanup("use apply with dict")

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 7]
```diff
@@ -690,585 +650,546 @@ public class Sched extends SchedV2 {
...
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONArray delays = conf.optJSONArray("delays");
+        val oconf = col.decks.confForDid(card.oDid)
+        var delays = conf.optJSONArray("delays")
         if (delays == null) {
-            delays = oconf.getJSONObject("lapse").getJSONArray("delays");
+            delays = oconf.getJSONObject("lapse").getJSONArray("delays")
         }
-        JSONObject dict = new JSONObject();
+        val dict = JSONObject()
```
Comment (lukstbit) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:982]: @KotlinCleanup("use apply with dict")

**Review by lukstbit — CHANGES_REQUESTED** [review]
Looks good, just some minor cleanups.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 4 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:785]: IDE lint

**Review by lukstbit — APPROVED** [review]

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 8]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
-        List<Long> all = new ArrayList<>(getCol().getDecks().children(did).values());
-        all.add(did);
-        unburyCardsForDeck(all);
+    override fun unburyCardsForDeck(did: Long) {
+        val all: MutableList<Long> = ArrayList(col.decks.children(did).values)
+        all.add(did)
+        unburyCardsForDeck(all)
     }
 
     /* Need to override. Otherwise it get SchedV2.mName variable*/
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1167]: Kotlin cleanup: don't need "get"

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 9]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
      * ***********************************************************
      * The methods below are not in LibAnki.
      * ***********************************************************
      */
-    @Override
-    public boolean haveBuried(long did) {
-        List<Long> all = new ArrayList<>(getCol().getDecks().children(did).values());
-        all.add(did);
-        return haveBuried(all);
+    override fun haveBuried(did: Long): Boolean {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1155]: Kotlin cleanup maybe: transform into an expression.
This and below

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 10]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
     /**
      * Deck finished state ******************************************************
      * *****************************************
      */
-
-
-    @Override
-    public boolean haveBuried() {
-        return haveBuried(getCol().getDecks().active());
+    override fun haveBuried(): Boolean {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1036]: KotlinCleanup: remove ": Boolean {"

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 11]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
-
-    private boolean _resched(@NonNull Card card) {
-        DeckConfig conf = _cardConf(card);
-        if (conf.getInt("dyn") == DECK_STD) {
-            return true;
-        }
-        return conf.getBoolean("resched");
+    private fun _resched(card: Card): Boolean {
+        val conf = _cardConf(card)
+        return if (conf.getInt("dyn") == DECK_STD) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1001]: Kotlin cleanup: `conf.getInt("dyn") == DECK_STD or conf.getBoolean("resched")`

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 12]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
+    override fun _lapseConf(card: Card): JSONObject {
+        val conf = _cardConf(card)
+        if (!card.isInDynamicDeck) {
+            return conf.getJSONObject("lapse")
         }
         // dynamic deck; override some attributes, use original deck for others
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONArray delays = conf.optJSONArray("delays");
+        val oconf = col.decks.confForDid(card.oDid)
+        var delays = conf.optJSONArray("delays")
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:982]: KotlinCleanup: uses ?:

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 13]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
+        val conf = _cardConf(card)
+        if (!card.isInDynamicDeck) {
+            return conf.getJSONObject("new")
         }
         // dynamic deck; override some attributes, use original deck for others
-        DeckConfig oconf = getCol().getDecks().confForDid(card.getODid());
-        JSONArray delays = conf.optJSONArray("delays");
+        val oconf = col.decks.confForDid(card.oDid)
+        var delays = conf.optJSONArray("delays")
         if (delays == null) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:958]: cleanup: uses ?:

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 14]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
+    /** Leech handler. True if card was a leech.  */
+    override fun _checkLeech(card: Card, conf: JSONObject): Boolean {
+        val lf = conf.getInt("leechFails")
         if (lf == 0) {
-            return false;
+            return false
         }
         // if over threshold or every half threshold reps after that
-        if (card.getLapses() >= lf && (card.getLapses() - lf) % Math.max(lf / 2, 1) == 0) {
+        if (card.lapses >= lf && (card.lapses - lf) % Math.max(lf / 2, 1) == 0) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:919]: KotlinCleanup: uses `card.apply` or even define it as a member of Card. `Card._checkLeech(conf: JSONObject)`

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 15]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
-        int order = terms.getInt(2);
-        SortOrder orderLimit = new SortOrder.AfterSqlOrderBy(_dynOrder(order, limit));
-        if (!TextUtils.isEmpty(search.trim())) {
-            search = String.format(Locale.US, "(%s)", search);
-        }
-        search = String.format(Locale.US, "%s -is:suspended -is:buried -deck:filtered -is:learn", search);
-        List<Long> ids = getCol().findCards(search, orderLimit);
+    private fun _fillDyn(deck: Deck): List<Long> {
+        val terms = deck.getJSONArray("terms").getJSONArray(0)
+        var search = terms.getString(0)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:848]: KotlinCleanup: transform this into val

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 16]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
-                    conf.getInt("maxIvl")));
-        } catch (JSONException e) {
-            throw new RuntimeException(e);
+            val idealIvl = _nextRevIvl(card, ease)
+            val conf = _revConf(card)
+            card.ivl = Math.min(
+                Math.max(_adjRevIvl(idealIvl), card.ivl + 1),
+                conf.getInt("maxIvl")
+            )
+        } catch (e: JSONException) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:812]: KotlinCleanup:  remove this catch/throw.
It was only useful to ensure we don't need to explicitly throw it in the method's declaration. That's not usefuly anymore

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 17]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
+        val fct = card.factor / 1000.0
+        val ivl2 =
+            _constrainedIvl(((card.ivl + delay / 4) * 1.2).toInt(), conf, card.ivl.toDouble())
+        val ivl3 = _constrainedIvl(((card.ivl + delay / 2) * fct).toInt(), conf, ivl2.toDouble())
+        val ivl4 = _constrainedIvl(
+            ((card.ivl + delay) * fct * conf.getDouble("ease4")).toInt(),
+            conf,
+            ivl3.toDouble()
+        )
         if (ease == Consts.BUTTON_TWO) {
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:787]: KotlinCleanup: do `val interval = ` with a last `else { 0}`

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 18]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
+    private fun _nextLapseIvl(card: Card, conf: JSONObject): Int {
+        return Math.max(conf.getInt("minInt"), (card.ivl * conf.getDouble("mult")).toInt())
     }
 
-
-    private void _rescheduleRev(@NonNull Card card, @Consts.BUTTON_TYPE int ease) {
+    private fun _rescheduleRev(card: Card, @BUTTON_TYPE ease: Int) {
         // update interval
-        card.setLastIvl(card.getIvl());
+        card.lastIvl = card.ivl
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:752]: KotlinCleanup: Either `card.apply` or even define it as `Card._rescheduleRev`.
Same for other methods rescheduling cards

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 19]
```diff
@@ -690,585 +652,549 @@ public class Sched extends SchedV2 {
...
                         // Note: libanki reverses mRevQueue and returns the last element in _getRevCard().
                         // AnkiDroid differs by leaving the queue intact and returning the *first* element
                         // in _getRevCard().
                     } else {
-                        Random r = new Random();
-                        r.setSeed(mToday);
-                        mRevQueue.shuffle(r);
+                        val r = Random()
+                        r.setSeed(mToday!!.toLong())
+                        mRevQueue.shuffle(r)
```
Comment (Arthur-Milchior) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:667]: KotlinCleanup: `mRevQueue(Random().apply(setSet(mToday!!.toLong()))`
 (same in schedv2)

**Review by Arthur-Milchior — COMMENTED** [review]
I probably missed some cleanup, this class is incredibly long. (And I plead guilty for part of it)
Still, I believe a bunch of cleanup that may improve code eventually

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 20]
```diff
@@ -16,591 +16,545 @@
...
     }
 
-
-    @Override
-    public @NonNull Counts counts(@NonNull Card card) {
-        Counts counts = counts();
-        Counts.Queue idx = countIdx(card);
-        if (idx == LRN) {
-            counts.addLrn(card.getLeft() / 1000);
+    override fun counts(card: Card): Counts {
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:100]: I don't think this is an improvement

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 21]
```diff
@@ -16,591 +16,545 @@
...
-            int rlim = _deckRevLimitSingle(deck, false);
+            var nlim = _deckNewLimitSingle(deck, false)
+            var rlim = _deckRevLimitSingle(deck, false)
             if (!TextUtils.isEmpty(p)) {
-                Integer[] parentLims = lims.get(Decks.normalizeName(p));
+                val parentLims = lims[Decks.normalizeName(p)]
                 // 'temporary for diagnosis of bug #6383'
-                Assert.that(parentLims != null, "Deck %s is supposed to have parent %s. It has not be found.", deckName, p);
-                nlim = Math.min(nlim, parentLims[0]);
+                that(
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:173]: This was an issue in the conversion. Fixed.

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 22]
```diff
@@ -16,591 +16,545 @@
...
+            val deckName = deck.getString("name")
+            val p = Decks.parent(deckName)
             // new
-            int nlim = _deckNewLimitSingle(deck, false);
-            int rlim = _deckRevLimitSingle(deck, false);
+            var nlim = _deckNewLimitSingle(deck, false)
+            var rlim = _deckRevLimitSingle(deck, false)
             if (!TextUtils.isEmpty(p)) {
-                Integer[] parentLims = lims.get(Decks.normalizeName(p));
+                val parentLims = lims[Decks.normalizeName(p)]
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:171]: This breaks the `Assert` below

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 23]
```diff
@@ -16,591 +16,545 @@
...
-
     /**
      * Return the next due card, or null.
      */
-    @Override
-    protected @Nullable Card _getCard() {
+    override fun _getCard(): Card? {
         // learning card due?
-        @Nullable Card c = _getLrnCard(false);
+        var c = _getLrnCard(false)
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:214]: This matched libAnki

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt [hunk 24]
```diff
@@ -16,591 +16,545 @@
...
-
-    @Override
-    protected void _rescheduleAsRev(@NonNull Card card, @NonNull JSONObject conf, boolean early) {
-        boolean lapse = (card.getType() == Consts.CARD_TYPE_REV);
+    override fun _rescheduleAsRev(card: Card, conf: JSONObject, early: Boolean) {
+        val lapse = card.type == Consts.CARD_TYPE_REV
         if (lapse) {
             if (_resched(card)) {
-                card.setDue(Math.max(mToday + 1, card.getODue()));
+                card.due = Math.max((mToday!! + 1).toLong(), card.oDue)
```
[the remark this answers is not in the document]
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:433]: This matched libAnki

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 10 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1036]: Moving this to an expression does this

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 14 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:919]: I disagree that this makes the code bettter

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 15 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:848]: This breaks `libAnki` compat

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 17 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:787]: this is handled in the IDE lint

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 18 above)
Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:752]: This breaks `libAnki` compat

**Comment by dae** [comment]
I don't recommend expending much effort on this file, since reviewing with the v1 scheduler is no longer possible since https://github.com/ankidroid/Anki-Android/pull/11808

**Comment by david-allison** [comment]
Agreed, just want to get to 0% Java so we can cut down the excess code and hopefully reduce compile times

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 20 above)
[the remark this answers is not in the document]
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:97]: ok

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt (hunk 10 above)
Comment (Arthur-Milchior) [reply AnkiDroid/src/main/java/com/ichi2/libanki/sched/Sched.kt:1040]: Yes, that what I meant. `haveBuried` can be replaced by an expression directly eventually

**Comment by Arthur-Milchior** [comment]
Merged. If we don't care about cleanup because the code will disappear, no need to wait for the cleanup

**Pull request merged**